### PR TITLE
Add base `Converter` class

### LINKED ISSUE #7392: Custom non-function converter
Chainer has introduced the `@chainer.dataset.converter` decorator for supporting `Device` specifiers in user-defined custom converters.

The problem is that it can only be applied to a function, because it creates a hidden attribute in the subject (`setattr(object(), ...)` causes an error).

Maybe we should provide a base class for converters.

(Reported in #7328.)

### PR DESCRIPTION
**Pull request by niboshi**
Fixes  #7392

### CODE REVIEW

**Comment by niboshi** [comment]
I'll write unit tests if this implementation is OK.

**Comment by emcastillo** [comment]
I think the solution is very elegant!

**Comment by emcastillo** [comment]
@niboshi do you plan to add the tests on this PR or should  I test and merge it :)

**Comment by niboshi** [comment]
@emcastillo I'll write tests. Please wait for a moment.

**Comment by niboshi** [comment]
@emcastillo Added tests.

**Comment by emcastillo** [comment]
LGTM

**Comment by niboshi** [comment]
TODO: update documentation

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Wrote docs.
I want to have the PR reviewed from @beam2d as well.

File: chainer/dataset/convert.py [hunk 1]
```diff
@@ -9,11 +8,77 @@
...
 
 
+class Converter(object):
+
+    """Base class of converters.
+
+    Converters receive batched data retrieved from iterators and perform
+    arbitrary transforms as well as device transfer.
+
+    Converters are required have a ``__call__`` method.
```
Comment (beam2d) [inline chainer/dataset/convert.py]:
```suggestion
    Implementation should override the ``__call__`` method.
```

**Comment by niboshi** [comment]
Thank you for the review.
@emcastillo PTAL.

**Comment by emcastillo** [comment]
LGTM

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
CI unrelated.

**Pull request merged**